### PR TITLE
[SVS] Add multi-vector support to the Tiered SVS Index

### PR DESCRIPTION
**Pull request by rfsaliev**
**Describe the changes in the pull request**

This PR adds multi-vector support to the Tiered SVS Index.

**NOTE: The pool request is published as 'Draft' because it is based on #690 and cannot be simply merged yet**

**Which issues this PR fixes**
1. #...
2. MOD...

**Main objects this PR modified**
1. ...
2. ...

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 1]
```diff
@@ -528,60 +533,69 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
...
+                assert(oldId == labels_to_move.size() - 1 &&
+                       "Old ID is not at the end of labels_to_move vector");
+
+                // remove the old label from labels_to_move at position oldId
+                labels_to_move.erase(labels_to_move.begin() + oldId);
+                if (newId != oldId) { // swap mode
+                    // Update the label at newId position
+                    labels_to_move[newId] = label;
+                }
             }
```
Comment (rfsaliev) [inline src/VecSim/algorithms/svs/svs_tiered.h:590]: @alonre24, can you please review this logic and `deleteVector()` as well?
Thank you.

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 2]
```diff
@@ -700,18 +714,53 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
...
+                    }
+                }
+
+                // After processing all deleting_ids, updated_ids should be empty in debug
+                // compilation.
+                assert(updated_ids.empty() &&
+                       "updated_ids should be empty after processing all deleting_ids");
+
+                ret = deleting_ids.size();
+                assert(ret >= 1 && "unexpected flat index deleteVector result");
```
Comment (rfsaliev) [inline src/VecSim/algorithms/svs/svs_tiered.h:757]: @alonre24, can you please review this logic and `updateSVSIndex()` as well?
Thank you.

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 3]
```diff
@@ -198,17 +199,16 @@ template <typename DataType>
...
 
-    // Add: true, Delete: false
-    using journal_record = std::pair<labelType, bool>;
+    // label, oldId, newId
+    using swap_record = std::tuple<labelType, idType, idType>;
     size_t trainingTriggerThreshold;
     size_t updateTriggerThreshold;
     size_t updateJobWaitTime;
-    std::vector<journal_record> journal;
-    std::shared_mutex journal_mutex;
```
Comment (rfsaliev) [inline src/VecSim/algorithms/svs/svs_tiered.h:211]: `journal_mutex` is not needed anymore because there is no any dedicated journal locks happen due to `TakeSnapshot()` removal.

**Review by rfsaliev — COMMENTED** [review]
Added some comments with explanation of changes.

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 4]
```diff
@@ -528,60 +533,71 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
...
+                auto data = flat_index->getDataByInternalId(i);
+                vectors_to_move.insert(vectors_to_move.end(), data, data + dim);
             }
+            // reset journal to the current frontend index state
+            swaps_journal.clear();
         } // release frontend index
 
         { // lock backend index for writing and add vectors there
-            std::scoped_lock lock(this->mainIndexGuard);
+            std::lock_guard lock(this->mainIndexGuard);
```
Comment (alonre24) [inline src/VecSim/algorithms/svs/svs_tiered.h:558]: Why change the lock type?

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 5]
```diff
@@ -528,60 +533,71 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
...
         }
         // clean-up frontend index
         { // lock frontend index for writing and delete moved vectors
-            std::scoped_lock lock(this->flatIndexGuard, this->journal_mutex);
-            // avoid deleting vectors from frontend which are updated/modified in the meantime
-            for (auto &p : this->journal) {
-                to_add.erase(p.first);
+            std::lock_guard lock(this->flatIndexGuard);
+
+            // Enumerate journal and swap label in labels_to_move to newId position.
```
Comment (alonre24) [inline src/VecSim/algorithms/svs/svs_tiered.h:568]: `swaps_journal` was cleared before we started to move vectors to the backend index. So here we are iterating the vector inserted into the flat index in the meantime, right?

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 6]
```diff
@@ -528,60 +533,71 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
...
+            for (const auto &p : this->swaps_journal) {
+                auto label = std::get<0>(p);
+                auto oldId = std::get<1>(p);
+                auto newId = std::get<2>(p);
+                if (oldId >= labels_to_move.size()) {
+                    continue; // Skip if oldId is out of bounds
```
Comment (alonre24) [inline src/VecSim/algorithms/svs/svs_tiered.h:574]: Has this new logic not been tested at all?

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 7]
```diff
@@ -700,18 +716,53 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
...
+                // Record swaps in the journal.
+                for (auto id : deleting_ids) {
+                    auto it = updated_ids.find(id);
+                    if (it != updated_ids.end()) {
+                        assert(id == it->first &&
+                               "id in updated_ids should match the id in deleting_ids");
+                        auto newId = id;
+                        auto oldId = it->second.first;
+                        auto oldLabel = it->second.second;
+                        this->swaps_journal.emplace_back(oldLabel, oldId, newId);
```
Comment (alonre24) [inline src/VecSim/algorithms/svs/svs_tiered.h:743]: Let's try to cover this edge case with a unit test, I believe we have a similar test in HNSW

File: src/VecSim/algorithms/svs/svs_tiered.h (hunk 3 above)
Comment (alonre24) [reply src/VecSim/algorithms/svs/svs_tiered.h:211]: Can you explain and document the purpose of the `swaps_journal`?

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 8]
```diff
@@ -528,60 +533,71 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
...
+                auto label = std::get<0>(p);
+                auto oldId = std::get<1>(p);
+                auto newId = std::get<2>(p);
+                if (oldId >= labels_to_move.size()) {
+                    continue; // Skip if oldId is out of bounds
+                }
+                assert(label == labels_to_move[oldId] &&
+                       "Journal label does not match the label in labels_to_move");
+
+                // TODO: verify the assumption asserted below
```
Comment (alonre24) [inline src/VecSim/algorithms/svs/svs_tiered.h:579]: TODO?

File: src/VecSim/algorithms/svs/svs_tiered.h (hunk 2 above)
Comment (alonre24) [reply src/VecSim/algorithms/svs/svs_tiered.h:759]: Reviewed, I believe it should work well :)

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 9]
```diff
@@ -665,18 +678,21 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
...
+        {
+            std::shared_lock lock(this->mainIndexGuard);
             update_threshold = this->backendIndex->indexSize() == 0 ? this->trainingTriggerThreshold
                                                                     : this->updateTriggerThreshold;
         }
-        { // Add vector to the frontend index and journal.
-            std::scoped_lock lock(this->flatIndexGuard, this->journal_mutex);
+        { // Add vector to the frontend index.
+            std::lock_guard lock(this->flatIndexGuard);
             ret = std::max(ret + this->frontendIndex->addVector(blob, label), 0);
```
Comment (alonre24) [inline src/VecSim/algorithms/svs/svs_tiered.h:695]: What should happen if we overwrite an existing label here? Should we account for it in the `swap_journal`?

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 10]
```diff
@@ -700,18 +716,53 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
         }();
 
         if (label_exists) {
-            std::scoped_lock lock(this->flatIndexGuard, this->journal_mutex);
-            if (this->frontendIndex->isLabelExists(label)) {
-                ret = this->frontendIndex->deleteVector(label);
-                assert(ret == 1 && "unexpected deleteVector result");
-                journal.emplace_back(label, false);
+            std::lock_guard lock(this->flatIndexGuard);
```
Comment (alonre24) [inline src/VecSim/algorithms/svs/svs_tiered.h:719]: Consider moving this to a helper function

**Review by alonre24 — COMMENTED** [review]
Nice! I reviewed the code, still left to review the tests

**Review by alonre24 — COMMENTED** [review]
reviewed unit tests as well

File: src/VecSim/algorithms/svs/svs_tiered.h (hunk 7 above)
Comment (rfsaliev) [reply src/VecSim/algorithms/svs/svs_tiered.h:805]: Added tests in test_svs_tiered.cpp called `testSwapJournalSingle` and `testSwapJournalMulti`.
Tried to emulate as worst cases as possible.

File: tests/unit/test_svs_tiered.cpp [hunk 11]
```diff
@@ -317,12 +333,26 @@ TYPED_TEST(SVSTieredIndexTest, addVector) {
     auto actual_mem = tiered_index->getAllocationSize();
     ASSERT_GE(expected_mem * 1.02, tiered_index->getAllocationSize());
     ASSERT_LE(expected_mem, tiered_index->getAllocationSize());
+
+    if constexpr (TypeParam::isMulti()) {
+        // Add another vector under the same label (create another insert job)
```
[the remark this answers is not in the document]
Comment (rfsaliev) [reply tests/unit/test_svs_tiered.cpp]: Done.

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 12]
```diff
@@ -529,60 +562,97 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
...
+                // Labels should be moved or removed from the end of index
+                assert(oldId == labels_to_move.size() - 1 &&
+                       "Old ID is not at the end of labels_to_move vector");
+
+                // remove the old label from labels_to_move at position oldId
+                labels_to_move.erase(labels_to_move.begin() + oldId);
+                if (newId != oldId) { // swap mode
+                    // Update the label at newId position
+                    labels_to_move[newId] = label;
+                }
```
[the remark this answers is not in the document]
Comment (rfsaliev) [reply src/VecSim/algorithms/svs/svs_tiered.h]: done.

File: tests/unit/test_svs_tiered.cpp [hunk 13]
```diff
@@ -2488,6 +2838,163 @@ TYPED_TEST(SVSTieredIndexTestBasic, switchDeleteModes) {
...
+
+        // Wait for the update job to start.
+        std::unique_lock lock(mtx);
+        EXPECT_TRUE(cv.wait_for(lock, std::chrono::seconds(100), [&] { return added_to_svs; }));
+
+        // update job paused, we have vectors 0-(n-1) in the index, let's do index modifications
+
+        // For single-value index, we expect to override the vector.
+        // For multi-value index, we expect to add a new vector.
+        const int update_count = is_multi ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (rfsaliev) [reply tests/unit/test_svs_tiered.cpp]: Thank you.
Now there are 2 dedicated tests: `testSwapJournalSingle` and `testSwapJournalMulti`

File: src/VecSim/algorithms/svs/svs_tiered.h [hunk 14]
```diff
@@ -666,18 +724,28 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
...
-            std::scoped_lock lock(this->flatIndexGuard, this->journal_mutex);
-            ret = std::max(ret + this->frontendIndex->addVector(blob, label), 0);
-            journal.emplace_back(label, true);
+        { // Add vector to the frontend index.
+            std::lock_guard lock(this->flatIndexGuard);
+            const auto ft_ret = this->frontendIndex->addVector(blob, label);
+
+            if (ft_ret == 0) { // Vector was overriden - add 'skiping' swap to the journal.
+                for (auto id : this->frontendIndex->getElementIds(label)) {
+                    this->swaps_journal.emplace_back(SKIP_LABEL, id, id);
```
[the remark this answers is not in the document]
Comment (rfsaliev) [reply src/VecSim/algorithms/svs/svs_tiered.h]: Added an assert.
